File: fsharp/pascals-triangle/PascalsTriangle.fs
module PascalsTriangle


let rows = function
    | 0 -> []
    | n ->
        let nextRow curRow =
            (0 :: curRow) @ [ 0 ]
            |> List.windowed 2
            |> List.map List.sum

        [ 1..n - 1 ]
        |> List.scan (fun acc _ -> nextRow acc) [ 1 ]


let rows' n =
    let createRow i =
        [ 1..i - 1 ]
        |> List.scan (fun acc j -> acc * (i - j) / j) 1

    [ 1..n ] |> List.map createRow

File: csharp/pascals-triangle/PascalsTriangle.cs
using System.Collections.Generic;
using System.Linq;

public static class PascalsTriangle
{
  public static IEnumerable<IEnumerable<int>> Calculate(int rows)
  {
    if (rows == 0) return Enumerable.Empty<IEnumerable<int>>();

    IEnumerable<IEnumerable<int>> firstRow = new[] { new[] { 1 } };

    return Enumerable
      .Range(1, rows - 1)
      .Aggregate(
        firstRow,
        (acc, _) => acc.Append(BuildNextRow(acc.Last()))
      );
  }

  private static IEnumerable<int> BuildNextRow(IEnumerable<int> row)
  {
    row = row.Prepend(0).Append(0);

    return row
      .Zip(row.Skip(1), (a, b) => (a, b))
      .Select(pair => pair.a + pair.b);
  }
}